src/SAWScript/MGU.hs

```diff
@@ -67,44 +67,118 @@ listSubst = Subst . M.fromList
 
 -- Most General Unifier {{{
 
-failMGU :: String -> Either String a
-failMGU = Left
-
-assert :: Bool -> String -> Either String ()
-assert b msg = unless b $ failMGU msg
-
-mgu :: LName -> Type -> Type -> Either String Subst
-mgu m (LType _ t) t2 = mgu m t t2
-mgu m t1 (LType _ t) = mgu m t1 t
-mgu m (TyUnifyVar i) t2 = bindVar m i t2
-mgu m t1 (TyUnifyVar i) = bindVar m i t1
-mgu m r1@(TyRecord ts1) r2@(TyRecord ts2) = do
-  assert (M.keys ts1 == M.keys ts2) $
-    "Record field type mismatch. Expected " ++ pShow r1 ++ " but got " ++ pShow r2 ++ " at " ++ show m
-  mgus m (M.elems ts1) (M.elems ts2)
-mgu m (TyCon tc1 ts1) (TyCon tc2 ts2)
-  | tc1 == tc2 = (mgus m ts1 ts2)
+{-
+ - If we fail here, the caller attaches the message we return to one
+ - of its own that gives the position of the best enclosing expression
+ - position available. It's important to make it clear that this is
+ - all one message. Since what we send back is printed on its own
+ - lines, indent it four spaces (the caller's message is already
+ - indented two) to make it more clearly part of its preceding text.
+ -
+ - We're going to print the expected and found types on their own
+ - lines because they can be large. (If so they can still be fairly
+ - illegible, but at least the user doesn't have to hunt for "found"
+ - in the middle of a multi-line print.) Also pad the prefix of the
+ - prints so that the types line up; this is helpful for longer types
+ - that still fit on one output line.
+ -
+ - We have recursed into the types and so if we get a mismatch we
+ - should print those types (this is important if they are e.g.
+ - elements in a large system of nested records and tuples) but also
+ - print the enclosing types on the way out. This can become rather
+ - verbose but makes it a lot easier to deal with errors deep within
+ - complex structures.
+ -
+ - The failure result is a start line, then all the expected/found
+ - lines we've accumulated, and finally, the in the middle, and the
+ - last seen expected/found lines arising from a function type, which
+ - we don't commit to using until we see a non-function type or get to
+ - the end. The idea is that if we pass through a chain of function
+ - types, each having the curried next argument to the previous, we
+ - only print the last and largest one.
+ -
+ - Note that although we append to the end of the expected/found list,
+ - we don't stick the start line in that list, because I keep going
+ - back and forth on whether the larger types should be printed first
+ - (prepending in failMGUadd) or last (appending). If we commit to
+ - appending we don't need to keep the start line separate.
+ -}
+
+type FailMGU = (String, [String], Maybe [String])
```

Comment:
I propose turning this tuple into a custom data type with named records, as this will make it easier to tell at a glance what each field does. I'd also appreciate having comments attached to each field, since it's not obvious to me from the description above what parts of the comment apply to which fields.

Reply:
I reworked the description above. Didn't help that the version you reviewed had at least one garbled sentence :-(

I also gave the type a data constructor, which is definitely an improvement in clarity. I tried using record syntax on top of that, but it isn't better. (LMK if you want to see the diff for that.) I could also change it to accumulate a list of pairs of types instead of message strings (and print the types at the end) -- that is probably in fact the right thing to do but I suspect this is all going to get ripped out later anyway.

(above is as of b82cffeefa693402e8fb3c45bb06e00b26604fe3)

Reply:
https://github.com/GaloisInc/saw-script/commit/b82cffeefa693402e8fb3c45bb06e00b26604fe3 is definitely an improvement.

I'm don't care too strongly about using record syntax or not, but I do think it would be nice to have a brief comment next to each field that describes what each field means. At the moment, you have to read the commentary written above `FailMGU` in order to know what each `String` means, but it would be nice to be able to tell at a glance what each field does.

Reply:
Whoops, almost forgot this. Added the comments in 78a675dc9c4ed2c3904a33769b892f36414d9230